### PR TITLE
UPSTREAM: <carry>: add validation for ingress.config.openshift.io component routes

### PR DESCRIPTION
**Pull request by deads2k**
/hold

just a POC, I don't plan to finish it out, but everything is stubbed for network edge.

/assign @candita

### CODE REVIEW

**Comment by deads2k** [comment]
I went ahead and finished this off.  Nobody wants to stop better API validation.  We just need to be sure that we don't leave non-updatable objects in our wake.

/hold cancel
/assign @Miciah @candita

File: openshift-kube-apiserver/admission/customresourcevalidation/ingress/validate_ingress.go [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+)
+
+const PluginName = "config.openshift.io/ValidateIngress"
+
+// Register registers a plugin
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName, func(config io.Reader) (admission.Interface, error) {
+		return customresourcevalidation.NewValidator(
+			map[schema.GroupResource]bool{
+				configv1.Resource("Ingresses"): true,
```
Comment (Miciah) [inline openshift-kube-apiserver/admission/customresourcevalidation/ingress/validate_ingress.go:25]: Maybe it doesn't matter, but shouldn't the resource name be lower-case?
```suggestion
				configv1.Resource("ingresses"): true,
```

File: openshift-kube-apiserver/admission/customresourcevalidation/ingress/validate_ingress.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+	if !ok {
+		return nil, append(allErrs,
+			field.NotSupported(field.NewPath("kind"), fmt.Sprintf("%T", uncastObj), []string{"Ingress"}),
+			field.NotSupported(field.NewPath("apiVersion"), fmt.Sprintf("%T", uncastObj), []string{"config.openshift.io/v1"}))
+	}
+
+	return obj, nil
+}
+
+type featureGateV1 struct {
```
Comment (Miciah) [inline openshift-kube-apiserver/admission/customresourcevalidation/ingress/validate_ingress.go:50]: Copypasta, or is that name intentional?

File: openshift-kube-apiserver/admission/customresourcevalidation/ingress/validate_ingress.go [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+}
+
+var knownFeatureSets = sets.NewString(
+	"",
+	string(configv1.TechPreviewNoUpgrade),
+	string(configv1.CustomNoUpgrade),
+	string(configv1.IPv6DualStackNoUpgrade),
+	string(configv1.LatencySensitive),
+)
```
Comment (Miciah) [inline openshift-kube-apiserver/admission/customresourcevalidation/ingress/validate_ingress.go:59]: This is unused.

File: openshift-kube-apiserver/admission/customresourcevalidation/ingress/validate_ingress.go [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+	// on create, we use the tightest validation
+	for i, currRoute := range spec.ComponentRoutes {
+		if errs := routeapihelpers.ValidateHost(
+			string(currRoute.Hostname),
+			"",
+			field.NewPath("spec.componentRoutes").Index(i).Child("hostname"),
+		); len(errs) > 0 {
+			allErrs = append(allErrs, errs...)
+		}
```
Comment (Miciah) [inline openshift-kube-apiserver/admission/customresourcevalidation/ingress/validate_ingress.go:72]: The length check is superfluous.  
```suggestion
		allErrs = append(allErrs, routeapihelpers.ValidateHost(
			string(currRoute.Hostname),
			"",
			field.NewPath("spec.componentRoutes").Index(i).Child("hostname"),
		)...)
```

File: openshift-kube-apiserver/admission/customresourcevalidation/ingress/validate_ingress.go [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+		for _, oldRoute := range oldSpec.ComponentRoutes {
+			if oldRoute.Name == currRoute.Name && oldRoute.Namespace == currRoute.Namespace {
+				previousRouteHostName = oldRoute.Hostname
+				break
+			}
+		}
+		// we don't enforce new validation rules if the hostname has not changed
+		if previousRouteHostName == currRoute.Hostname {
+			continue
+		}
```
Comment (Miciah) [inline openshift-kube-apiserver/admission/customresourcevalidation/ingress/validate_ingress.go:105]: This logic allows the following test case to pass; is the test case valid, or should it expect an error?  
```go
		{
			name:        "no previous value to empty",
			hostname:    "",
			oldHostname: "nohost",
			noPrevious:  true,
			expectedErr: ``,
		},
```

Comment (deads2k) [reply openshift-kube-apiserver/admission/customresourcevalidation/ingress/validate_ingress.go:105]: very well found!

**Comment by deads2k** [comment]
Admission validation carry patches are accepted for types that must logically exist before their operators are present.

**Comment by deads2k** [comment]
diff of diffs is identical.  relabeling.

File: openshift-kube-apiserver/admission/customresourcevalidation/ingress/validate_ingress_test.go [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+		{
+			name:        "no change",
+			hostname:    "111!!! invalid I think",
+			oldHostname: "111!!! invalid I think",
+			expectedErr: ``,
+		},
+		{
+			name:        "change from invalid to valid",
+			hostname:    "host.com",
+			oldHostname: "host",
```
Comment (candita) [inline openshift-kube-apiserver/admission/customresourcevalidation/ingress/validate_ingress_test.go:82]: A single label hostname is a valid hostname and should not fail validation

File: openshift-kube-apiserver/admission/customresourcevalidation/ingress/validate_ingress_test.go [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+			expectedErr: ``,
+		},
+		{
+			name:        "change from valid to used-to-be-valid",
```
Comment (candita) [inline openshift-kube-apiserver/admission/customresourcevalidation/ingress/validate_ingress_test.go:86]: Both are valid.

**Comment by deads2k** [comment]
> /hold

We have to hold this until we can work out whether https://github.com/openshift/kubernetes/pull/1216#discussion_r829439243 is desired behavior or not.  I don't actually know.

cc @Miciah

**Comment by deads2k** [comment]
openshift/api PR was updated to loosen only.

/close

**Pull request closed without merging**